src/Components/test/E2ETest/ServerExecutionTests/InteropReliabilityTests.cs

```diff
@@ -0,0 +1,349 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Ignitor;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.E2ETesting;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETests.ServerExecutionTests
+{
+    public class InteropReliabilityTests : ServerTestBase<AspNetSiteServerFixture>
+    {
+        private const int DefaultLatencyTimeout = 500;
+
+        public InteropReliabilityTests(
+            BrowserFixture browserFixture,
+            AspNetSiteServerFixture serverFixture,
+            ITestOutputHelper output)
+            : base(browserFixture, serverFixture, output)
+        {
+            serverFixture.BuildWebHostMethod = TestServer.Program.BuildWebHost;
+        }
+
+        public BlazorClient Client { get; set; } = new BlazorClient();
+
+        public override Task InitializeAsync()
+        {
+            // Do nothing.
+            return Task.CompletedTask;
+        }
+
+        [Fact]
+        public async Task CannotInvokeNonJSInvokableMethods()
+        {
+            // Arrange
+            var expectedError = "[\"1\"," +
+                "false," +
+                "\"There was an exception invoking \\u0027WriteAllText\\u0027 on assembly \\u0027System.IO.FileSystem\\u0027. For more details turn on detailed exceptions in \\u0027CircuitOptions.JSInteropDetailedErrors\\u0027\"]";
+
+            var (interopCalls, batches) = ConfigureClient();
+            await GoToTestComponent(batches);
+
+            // Act
+            Client.InvokeDotNetMethod(
+                "1",
+                "System.IO.FileSystem",
+                "WriteAllText",
+                null,
+                JsonSerializer.Serialize(new[] { ".\\log.txt", "log" }));
+
+            await Task.Delay(DefaultLatencyTimeout);
```

Comment:
As we dicussed offline, this pattern needs to be replaced with something more resiliant, this will definitely be flaky.

Reply:
Fixed it. There still needs to be a timeout though, to avoid the tests from hanging, but I've switched to use a TCS all over the place.